packages/block-editor/src/components/off-canvas-editor/block-edit-button.js

```diff
@@ -2,26 +2,208 @@
+				'core/navigation-link',
+				{
+					id,
+					label: title.rendered,
+					url,
+					type,
+					kind: 'post-type',
+				},
+				innerBlocks
+			);
+
+			if ( ! parent ) {
+				navigationLinks.push( linkMap[ id ] );
+			} else {
+				if ( ! linkMap[ parent ] ) {
+					// Use a placeholder if the child appears before parent in list.
+					linkMap[ parent ] = { innerBlocks: [] };
+				}
+				const parentLinkInnerBlocks = linkMap[ parent ].innerBlocks;
+				parentLinkInnerBlocks.push( linkMap[ id ] );
+			}
+		} );
+
+		// Transform all links with innerBlocks into Submenus. This can't be done
+		// sooner because page objects have no information on their children.
+
+		const transformSubmenus = ( listOfLinks ) => {
+			listOfLinks.forEach( ( block, index, listOfLinksArray ) => {
+				const { attributes, innerBlocks } = block;
+				if ( innerBlocks.length !== 0 ) {
+					transformSubmenus( innerBlocks );
+					const transformedBlock = createBlock(
+						'core/navigation-submenu',
+						attributes,
+						innerBlocks
+					);
+					listOfLinksArray[ index ] = transformedBlock;
+				}
+			} );
+		};
+
+		transformSubmenus( navigationLinks );
+
+		replaceBlock( clientId, navigationLinks );
+	};
+
+const ConvertToLinksModal = ( { onClose, clientId, pages } ) => {
+	const hasPages = !! pages?.length;
+
+	const { replaceBlock } = useDispatch( blockEditorStore );
+
+	return (
+		<Modal
+			closeLabel={ __( 'Close' ) }
+			onRequestClose={ onClose }
+			title={ __( 'Customize this menu' ) }
+			className={ 'wp-block-page-list-modal' }
+			aria={ { describedby: 'wp-block-page-list-modal__description' } }
+		>
+			<p id={ 'wp-block-page-list-modal__description' }>
+				{ __(
+					'This menu is automatically kept in sync with pages on your site. You can manage the menu yourself by clicking customize below.'
+				) }
+			</p>
+			<div className="wp-block-page-list-modal-buttons">
+				<Button variant="tertiary" onClick={ onClose }>
+					{ __( 'Cancel' ) }
+				</Button>
+				<Button
+					variant="primary"
+					disabled={ ! hasPages }
+					onClick={ convertSelectedBlockToNavigationLinks( {
+						pages,
+						replaceBlock,
+						clientId,
+						createBlock: create,
+					} ) }
+				>
+					{ __( 'Customize' ) }
+				</Button>
+			</div>
+		</Modal>
+	);
+};
+
 export default forwardRef( function BlockEditButton(
 	{ clientId, ...props },
 	ref
 ) {
 	const { selectBlock } = useDispatch( blockEditorStore );
+	const [ convertModalOpen, setConvertModalOpen ] = useState( false );
+	const { pages, totalPages } = usePageData();
+
+	const block = useSelect(
+		( select ) => {
+			return select( blockEditorStore ).getBlock( clientId );
+		},
+		[ clientId ]
+	);
+
+	const allowConvertToLinks =
+		'core/page-list' === block.name && totalPages <= MAX_PAGE_COUNT;
 
 	const onClick = () => {
-		selectBlock( clientId );
+		if ( allowConvertToLinks ) {
+			setConvertModalOpen( ! convertModalOpen );
+		} else {
+			selectBlock( clientId );
+		}
```

Comment:
I don't think this component should know anything about this concept. Lets lift this up and pass down as a prop. Perhaps if the `onClick` prop is defined then call that, else default to `selectBlock`?